packages/sources/eth-beacon/src/transport/balance.ts

```diff
@@ -53,43 +54,70 @@ export type Address = {
+  provider!: ethers.providers.JsonRpcProvider
   requester!: Requester
+  config!: BalanceTransportTypes['Settings']
+  endpointName!: string
 
   async initialize(
     dependencies: TransportDependencies<BalanceTransportTypes>,
-    _adapterSettings: BalanceTransportTypes['Settings'],
-    _endpointName: string,
+    adapterSettings: BalanceTransportTypes['Settings'],
+    endpointName: string,
     transportName: string,
   ): Promise<void> {
-    this.responseCache = dependencies.responseCache
+    await super.initialize(dependencies, adapterSettings, endpointName, transportName)
+    this.config = adapterSettings
+    this.provider = new ethers.providers.JsonRpcProvider(
+      adapterSettings.ETH_EXECUTION_RPC_URL,
+      adapterSettings.CHAIN_ID,
+    )
     this.requester = dependencies.requester
-    this.name = transportName
+    this.endpointName = endpointName
   }
-  async foregroundExecute(
-    req: AdapterRequest<typeof inputParameters.validated>,
-    adapterSettings: BalanceTransportTypes['Settings'],
-  ): Promise<AdapterResponse<BalanceTransportTypes['Response']>> {
-    const params = req.requestContext.data
 
+  async backgroundHandler(context: EndpointContext<BaseEndpointTypes>, entries: RequestParams[]) {
+    if (!entries.length) {
+      await sleep(context.adapterSettings.BACKGROUND_EXECUTE_MS)
+      return
+    }
+    await Promise.all(entries.map(async (param) => this.handleRequest(param)))
+  }
+
+  async handleRequest(param: RequestParams) {
+    let response: AdapterResponse<BaseEndpointTypes['Response']>
+    try {
+      response = await this._handleRequest(param)
+    } catch (e) {
+      const errorMessage = e instanceof Error ? e.message : 'Unknown error occurred'
+      response = {
+        statusCode: 502,
+        errorMessage,
+        timestamps: {
+          providerDataRequestedUnixMs: 0,
+          providerDataReceivedUnixMs: 0,
+          providerIndicatedTimeUnixMs: undefined,
+        },
+      }
+    }
+    await this.responseCache.write(this.name, [{ params: param, response }])
+  }
+
+  async _handleRequest(
+    params: RequestParams,
+  ): Promise<AdapterResponse<BaseEndpointTypes['Response']>> {
     const url = `/eth/v1/beacon/states/${params.stateId}/validators`
     const statusList = params.validatorStatus?.join(',')
-    const batchSize = adapterSettings.BATCH_SIZE
+    const batchSize = this.config.BATCH_SIZE
     const responses = []
 
     // If adapter configured with 0 batch size, put all validators in one request to allow skipping batching
     if (batchSize === 0) {
       const addresses = params.addresses.map(({ address }) => address).join(',')
       responses.push(
-        await this.queryBeaconChain(
-          adapterSettings,
-          req.requestContext.endpointName,
-          url,
-          addresses,
-          statusList,
-        ),
+        await this.queryBeaconChain(this.config, this.endpointName, url, addresses, statusList),
```

Comment:
Could we just use `this.config` and `this.endpointName` directly in `queryBeaconChain`? It'll clear up the parameter clutter here